File: crates/acton-reactive/7.1.1/examples/ipc_client_libraries/python/example_client.py
#!/usr/bin/env python3
"""
Example client demonstrating the Acton IPC client library.

This example shows how to:
1. Connect to an acton-reactive IPC server
2. Send request-response messages
3. Use streaming responses
4. Subscribe to push notifications
5. Discover available services

Usage:
    # First, start the example server (from the Rust side):
    cargo run --example ipc_client_libraries_server

    # Then run this client:
    python example_client.py
"""

import asyncio
import sys
from pathlib import Path

# Import from the local library
from acton_ipc import (
    ActonIpcClient,
    ActonIpcClientSync,
    ConnectionError,
    PushNotification,
    ServerError,
    TimeoutError,
    get_default_socket_path,
    socket_exists,
)


async def demo_basic_request(client: ActonIpcClient):
    """Demonstrate basic request-response pattern."""
    print("\n=== Basic Request-Response ===")

    try:
        response = await client.request(
            target="calculator",
            message_type="Add",
            payload={"a": 42, "b": 8},
            timeout_ms=5000,
        )

        if response.success:
            print(f"  Result: {response.payload}")
        else:
            print(f"  Error: {response.error}")

    except ServerError as e:
        print(f"  Server error: {e} (code: {e.code})")
    except TimeoutError as e:
        print(f"  Timeout: {e}")


async def demo_fire_and_forget(client: ActonIpcClient):
    """Demonstrate fire-and-forget pattern."""
    print("\n=== Fire and Forget ===")

    await client.fire_and_forget(
        target="logger",
        message_type="LogEvent",
        payload={
            "level": "info",
            "message": "Hello from Python client!",
        },
    )
    print("  Message sent (no response expected)")


async def demo_streaming(client: ActonIpcClient):
    """Demonstrate streaming response pattern."""
    print("\n=== Streaming Response ===")

    try:
        frame_count = 0
        async for frame in client.stream(
            target="search",
            message_type="SearchQuery",
            payload={"query": "test", "limit": 5},
            timeout_ms=10000,
        ):
            frame_count += 1
            if frame.error:
                print(f"  Stream error: {frame.error}")
                break

            print(f"  Frame {frame.sequence}: {frame.payload}")

            if frame.is_final:
                print(f"  Stream complete ({frame_count} frames)")
                break

    except TimeoutError as e:
        print(f"  Stream timeout: {e}")


async def demo_discovery(client: ActonIpcClient):
    """Demonstrate service discovery."""
    print("\n=== Service Discovery ===")

    try:
        discovery = await client.discover()

        print(f"  Protocol: v{discovery.protocol_version.get('current', '?')}")
        print(f"  Capabilities: {discovery.protocol_version.get('capabilities', {})}")
        print(f"  Actors: {[a.get('name') for a in discovery.actors]}")
        print(f"  Message Types: {discovery.message_types}")

    except Exception as e:
        print(f"  Discovery failed: {e}")


async def demo_subscriptions(client: ActonIpcClient):
    """Demonstrate push notifications via subscriptions."""
    print("\n=== Subscriptions ===")

    received_notifications = []

    def on_notification(notification: PushNotification):
        received_notifications.append(notification)
        print(f"  [PUSH] {notification.message_type}: {notification.payload}")

    client.on_push(on_notification)

    # Subscribe to message types
    success = await client.subscribe(["PriceUpdate", "StatusChange"])
    print(f"  Subscription result: {'success' if success else 'failed'}")

    # In a real scenario, we would wait for push notifications here
    # For demo purposes, we'll just show that we're subscribed
    print("  Waiting for notifications (3 seconds)...")
    await asyncio.sleep(3)

    print(f"  Received {len(received_notifications)} notifications")

    # Unsubscribe
    await client.unsubscribe(["PriceUpdate", "StatusChange"])
    print("  Unsubscribed")


async def main_async():
    """Async main entry point."""
    # Determine socket path
    socket_path = get_default_socket_path("ipc_client_example")

    # Allow override via command line
    if len(sys.argv) > 1:
        socket_path = Path(sys.argv[1])

    print(f"Acton IPC Python Client Example")
    print(f"================================")
    print(f"Socket path: {socket_path}")

    # Check if socket exists
    if not socket_exists(socket_path):
        print(f"\nError: Socket not found at {socket_path}")
        print("\nMake sure the example server is running:")
        print("  cargo run --example ipc_client_libraries_server")
        sys.exit(1)

    # Connect and run demos
    try:
        async with ActonIpcClient(socket_path) as client:
            print("\nConnected to server!")

            await demo_discovery(client)
            await demo_basic_request(client)
            await demo_fire_and_forget(client)
            await demo_streaming(client)
            await demo_subscriptions(client)

            print("\n=== All demos complete ===")

    except ConnectionError as e:
        print(f"\nConnection error: {e}")
        sys.exit(1)


def main_sync():
    """Synchronous main entry point (alternative)."""
    socket_path = get_default_socket_path("ipc_client_example")

    if len(sys.argv) > 1:
        socket_path = Path(sys.argv[1])

    print(f"Acton IPC Python Client (Sync)")
    print(f"==============================")
    print(f"Socket path: {socket_path}")

    if not socket_exists(socket_path):
        print(f"\nError: Socket not found at {socket_path}")
        sys.exit(1)

    with ActonIpcClientSync(socket_path) as client:
        print("\nConnected!")

        # Basic request
        response = client.request(
            target="calculator",
            message_type="Add",
            payload={"a": 10, "b": 20},
        )
        print(f"Result: {response.payload}")


if __name__ == "__main__":
    # Use async version by default
    asyncio.run(main_async())

File: crates/acton-reactive/7.1.0/examples/ipc_client_libraries/nodejs/src/example-client.ts
#!/usr/bin/env npx ts-node
/**
 * Example client demonstrating the Acton IPC client library.
 *
 * This example shows how to:
 * 1. Connect to an acton-reactive IPC server
 * 2. Send request-response messages
 * 3. Use streaming responses
 * 4. Subscribe to push notifications
 * 5. Discover available services
 *
 * Usage:
 *   # First, start the example server (from the Rust side):
 *   cargo run --example ipc_client_libraries_server
 *
 *   # Then run this client:
 *   npm run example
 *   # or
 *   npx ts-node src/example-client.ts
 */

import {
  ActonIpcClient,
  getDefaultSocketPath,
  socketExists,
  PushNotification,
  ServerError,
  TimeoutError,
  ConnectionError,
} from "./index";

async function demoBasicRequest(client: ActonIpcClient): Promise<void> {
  console.log("\n=== Basic Request-Response ===");

  try {
    const response = await client.request(
      "calculator",
      "Add",
      { a: 42, b: 8 },
      5000,
    );

    if (response.success) {
      console.log(`  Result: ${JSON.stringify(response.payload)}`);
    } else {
      console.log(`  Error: ${response.error}`);
    }
  } catch (err) {
    if (err instanceof ServerError) {
      console.log(`  Server error: ${err.message} (code: ${err.code})`);
    } else if (err instanceof TimeoutError) {
      console.log(`  Timeout: ${err.message}`);
    } else {
      throw err;
    }
  }
}

async function demoFireAndForget(client: ActonIpcClient): Promise<void> {
  console.log("\n=== Fire and Forget ===");

  await client.fireAndForget("logger", "LogEvent", {
    level: "info",
    message: "Hello from Node.js client!",
  });

  console.log("  Message sent (no response expected)");
}

async function demoStreaming(client: ActonIpcClient): Promise<void> {
  console.log("\n=== Streaming Response ===");

  try {
    let frameCount = 0;

    for await (const frame of client.stream(
      "search",
      "SearchQuery",
      { query: "test", limit: 5 },
      10000,
    )) {
      frameCount++;

      if (frame.error) {
        console.log(`  Stream error: ${frame.error}`);
        break;
      }

      console.log(
        `  Frame ${frame.sequence}: ${JSON.stringify(frame.payload)}`,
      );

      if (frame.is_final) {
        console.log(`  Stream complete (${frameCount} frames)`);
        break;
      }
    }
  } catch (err) {
    if (err instanceof TimeoutError) {
      console.log(`  Stream timeout: ${err.message}`);
    } else {
      throw err;
    }
  }
}

async function demoDiscovery(client: ActonIpcClient): Promise<void> {
  console.log("\n=== Service Discovery ===");

  try {
    const discovery = await client.discover();

    console.log(`  Protocol: v${discovery.protocol_version?.current ?? "?"}`);
    console.log(
      `  Capabilities: ${JSON.stringify(discovery.protocol_version?.capabilities ?? {})}`,
    );
    console.log(
      `  Actors: ${discovery.actors?.map((a) => a.name).join(", ") || "none"}`,
    );
    console.log(
      `  Message Types: ${discovery.message_types?.join(", ") || "none"}`,
    );
  } catch (err) {
    console.log(`  Discovery failed: ${err}`);
  }
}

async function demoSubscriptions(client: ActonIpcClient): Promise<void> {
  console.log("\n=== Subscriptions ===");

  const receivedNotifications: PushNotification[] = [];

  client.onPush((notification) => {
    receivedNotifications.push(notification);
    console.log(
      `  [PUSH] ${notification.message_type}: ${JSON.stringify(notification.payload)}`,
    );
  });

  // Subscribe to message types
  const success = await client.subscribe(["PriceUpdate", "StatusChange"]);
  console.log(`  Subscription result: ${success ? "success" : "failed"}`);

  // Wait for notifications
  console.log("  Waiting for notifications (3 seconds)...");
  await new Promise((resolve) => setTimeout(resolve, 3000));

  console.log(`  Received ${receivedNotifications.length} notifications`);

  // Unsubscribe
  await client.unsubscribe(["PriceUpdate", "StatusChange"]);
  console.log("  Unsubscribed");
}

async function main(): Promise<void> {
  // Determine socket path
  let socketPath = getDefaultSocketPath("ipc_client_example");

  // Allow override via command line
  if (process.argv[2]) {
    socketPath = process.argv[2];
  }

  console.log("Acton IPC Node.js Client Example");
  console.log("=================================");
  console.log(`Socket path: ${socketPath}`);

  // Check if socket exists
  if (!socketExists(socketPath)) {
    console.log(`\nError: Socket not found at ${socketPath}`);
    console.log("\nMake sure the example server is running:");
    console.log("  cargo run --example ipc_client_libraries_server");
    process.exit(1);
  }

  // Connect and run demos
  const client = new ActonIpcClient(socketPath);

  try {
    await client.connect();
    console.log("\nConnected to server!");

    await demoDiscovery(client);
    await demoBasicRequest(client);
    await demoFireAndForget(client);
    await demoStreaming(client);
    await demoSubscriptions(client);

    console.log("\n=== All demos complete ===");
  } catch (err) {
    if (err instanceof ConnectionError) {
      console.log(`\nConnection error: ${err.message}`);
      process.exit(1);
    }
    throw err;
  } finally {
    await client.close();
  }
}

main().catch(console.error);
